File: src/lib/api/repair.ts
// Initiative 8 -- refurbishable spares tracking: the repair register, the duplicate
// guard, and condition-to-repair declarations.
import { apiFetch } from "@/lib/api/client"

/** One open repair chain: a unit that is out at a vendor (or waiting to go). */
export interface RepairChain {
  material_id: number
  material_code: string | null
  material_description: string | null
  material_group: string | null
  plant: string | null
  repair_pr_number: string | null
  repair_pr_id: number | null
  repair_po_number: string | null
  repair_po_id: number | null
  supplier_id: number | null
  vendor: string | null
  quantity_under_repair: number
  repair_value: number
  opened_at: string | null
  expected_return: string | null
  days_open: number | null
  days_overdue: number
  overdue: boolean
  stage: string | null
}

/** A register row: the chain, plus the inventory position beside it. */
export interface RepairRegisterRow extends RepairChain {
  stock_on_hand: number
  unit_of_measure: string | null
  reorder_point: number | null
  reorder_triggered: boolean
  criticality: string | null
  new_unit_cost: number
  new_lead_time_days: number
  declarations_pending: number
  declarations_complete: number
  duplicate_risk: boolean
}

export interface RepairRegisterSummary {
  open_chain_count: number
  total_quantity_under_repair: number
  total_value_under_repair: number
  overdue_count: number
  reorder_triggered_count: number
  duplicate_risk_count: number
  pending_declaration_count: number
  average_days_open: number | null
}

export interface RepairRegister {
  items: RepairRegisterRow[]
  summary: RepairRegisterSummary
  total: number
}

export interface EconomicEvaluation {
  material_id: number
  material_code: string | null
  material_description: string | null
  plant: string | null
  quantity: number
  repair_total_cost: number
  repair_cost_basis: string
  repair_reference: string | null
  repair_vendor: string | null
  repair_expected_return: string | null
  repair_days_until_return: number | null
  repair_is_overdue: boolean
  new_total_cost: number
  new_unit_cost: number
  new_lead_time_days: number
  saving_if_repair_used: number
  saving_pct: number | null
  repair_arrives_sooner: boolean
}

/** What the guard reports before a new-buy requisition is raised. Advisory only. */
export interface DuplicateCheck {
  material_id: number
  material_code: string | null
  material_description: string | null
  plant: string | null
  is_repairable: boolean
  has_active_chain: boolean
  total_quantity_under_repair: number
  earliest_expected_return: string | null
  chains: RepairChain[]
  economics: EconomicEvaluation | null
  attestation_required: boolean
  attestation_statement: string | null
}

export interface Attestation {
  id: number
  rr_id: number | null
  rr_number: string | null
  rr_status: string | null
  material_id: number | null
  material_code: string | null
  material_description: string | null
  plant: string | null
  origin: "MANUAL" | "MRP" | "CHAT" | null
  status: "COMPLETE" | "PENDING" | null
  statement: string | null
  declared_by: number | null
  declared_by_name: string | null
  declared_at: string | null
  created_at: string | null
  duplicate_flag: boolean
  chain_snapshot: DuplicateContext | null
}

export interface PendingDeclaration {
  attestation_id: number
  rr_id: number
  rr_number: string | null
  rr_status: string | null
  plant: string | null
  department: string | null
  priority: string | null
  trigger_type: string | null
  origin: string | null
  material_id: number | null
  material_code: string | null
  material_description: string | null
  requester: string | null
  created_at: string | null
  duplicate_flag: boolean
  chain_snapshot: DuplicateContext | null
}

/** The shape stored on a flagged RR/PR and shown to the approver. Written identically by
 * the data generator and by rr_service at runtime. */
export interface DuplicateContext {
  detected_at: string
  plant: string | null
  chain_count: number
  materials: {
    material_id: number
    material_code: string | null
    material_description: string | null
    plant: string | null
    total_quantity_under_repair: number
    earliest_expected_return: string | null
    chains: RepairChain[]
  }[]
}

export interface RegisterParams {
  plant?: string
  status?: "OVERDUE" | "IN_FLIGHT" | "REORDER_TRIGGERED"
  material_group?: string
  search?: string
  [key: string]: string | number | boolean | undefined
}

export function getRepairRegister(params: RegisterParams = {}): Promise<RepairRegister> {
  return apiFetch<RepairRegister>("/repair/register", { params })
}

export function getRepairPlants(): Promise<string[]> {
  return apiFetch<string[]>("/repair/register/plants")
}

export function checkRepairChain(materialId: number, plant?: string): Promise<DuplicateCheck> {
  return apiFetch<DuplicateCheck>("/repair/chain-check", {
    params: { material_id: materialId, plant },
  })
}

export function getRepairEconomics(materialId: number, plant?: string): Promise<EconomicEvaluation> {
  return apiFetch<EconomicEvaluation>("/repair/economics", {
    params: { material_id: materialId, plant },
  })
}

export function listAttestations(
  params: { status?: string; origin?: string; plant?: string } = {}
): Promise<Attestation[]> {
  return apiFetch<Attestation[]>("/repair/attestations", { params })
}

export function listPendingDeclarations(plant?: string): Promise<PendingDeclaration[]> {
  return apiFetch<PendingDeclaration[]>("/repair/attestations/pending", { params: { plant } })
}

export function declareAttestation(id: number, note?: string): Promise<Attestation> {
  return apiFetch<Attestation>(`/repair/attestations/${id}/declare`, {
    method: "POST",
    json: { note },
  })
}

File: backend/app/schemas/repair.py
"""Initiative 8 response/request shapes -- repair register, duplicate guard, attestations."""

from app.schemas.common import ORMBase


class RepairChainOut(ORMBase):
    material_id: int
    material_code: str | None = None
    material_description: str | None = None
    material_group: str | None = None
    plant: str | None = None
    repair_pr_number: str | None = None
    repair_pr_id: int | None = None
    repair_po_number: str | None = None
    repair_po_id: int | None = None
    supplier_id: int | None = None
    vendor: str | None = None
    quantity_under_repair: float
    repair_value: float
    opened_at: str | None = None
    expected_return: str | None = None
    days_open: int | None = None
    days_overdue: int = 0
    overdue: bool = False
    stage: str | None = None


class RegisterRowOut(RepairChainOut):
    stock_on_hand: int
    unit_of_measure: str | None = None
    reorder_point: int | None = None
    reorder_triggered: bool = False
    criticality: str | None = None
    new_unit_cost: float = 0.0
    new_lead_time_days: int = 0
    declarations_pending: int = 0
    declarations_complete: int = 0
    duplicate_risk: bool = False


class RegisterSummaryOut(ORMBase):
    open_chain_count: int
    total_quantity_under_repair: float
    total_value_under_repair: float
    overdue_count: int
    reorder_triggered_count: int
    duplicate_risk_count: int
    pending_declaration_count: int
    average_days_open: float | None = None


class RegisterOut(ORMBase):
    items: list[RegisterRowOut]
    summary: RegisterSummaryOut
    total: int


class DuplicateCheckOut(ORMBase):
    material_id: int
    material_code: str | None = None
    material_description: str | None = None
    plant: str | None = None
    is_repairable: bool
    has_active_chain: bool
    total_quantity_under_repair: float = 0.0
    earliest_expected_return: str | None = None
    chains: list[RepairChainOut] = []
    # Present only when a chain is active -- there is no decision to inform otherwise.
    economics: "EconomicEvaluationOut | None" = None
    attestation_required: bool = False
    attestation_statement: str | None = None


class EconomicEvaluationOut(ORMBase):
    material_id: int
    material_code: str | None = None
    material_description: str | None = None
    plant: str | None = None
    quantity: float
    repair_total_cost: float
    repair_cost_basis: str
    repair_reference: str | None = None
    repair_vendor: str | None = None
    repair_expected_return: str | None = None
    repair_days_until_return: int | None = None
    repair_is_overdue: bool = False
    new_total_cost: float
    new_unit_cost: float
    new_lead_time_days: int
    saving_if_repair_used: float
    saving_pct: float | None = None
    repair_arrives_sooner: bool = False


DuplicateCheckOut.model_rebuild()


class AttestationOut(ORMBase):
    id: int
    rr_id: int | None = None
    rr_number: str | None = None
    rr_status: str | None = None
    material_id: int | None = None
    material_code: str | None = None
    material_description: str | None = None
    plant: str | None = None
    origin: str | None = None
    status: str | None = None
    statement: str | None = None
    declared_by: int | None = None
    declared_by_name: str | None = None
    declared_at: str | None = None
    created_at: str | None = None
    duplicate_flag: bool = False
    chain_snapshot: dict | None = None


class PendingDeclarationOut(ORMBase):
    attestation_id: int
    rr_id: int
    rr_number: str | None = None
    rr_status: str | None = None
    plant: str | None = None
    department: str | None = None
    priority: str | None = None
    trigger_type: str | None = None
    origin: str | None = None
    material_id: int | None = None
    material_code: str | None = None
    material_description: str | None = None
    requester: str | None = None
    created_at: str | None = None
    duplicate_flag: bool = False
    chain_snapshot: dict | None = None


class DeclareRequest(ORMBase):
    note: str | None = None
